src/Core/Checkout/Cart/Cleanup/CleanupCartTaskHandler.php

```diff
@@ -34,7 +34,7 @@ public function run(): void
         $time->modify(sprintf('-%s day', $this->days));
 
         $this->connection->executeStatement(
-            'DELETE FROM cart WHERE created_at <= :timestamp',
+            'DELETE FROM cart WHERE COALESCE(updated_at, created_at) <= :timestamp',
```

Comment:
Wouldn't be better, performance-wise, to set `updated_at` to same value as `created_at` at insert, adding an index on column `updated_at` since now is missing, and use `WHERE updated_at <= :timestamp` which would use the new index?

Reply:
Hi @mynameisbogdan

- Not intending to set updated_at aand created_at together because no other entity in Shopware 6 does so, there really is no sense to break the pattern here.
- I have adjusted the `WHERE` clause so in the future it could take advantage of indices, if necessary. This should be enough in case indices are added later on.
- The `cart` table doesn't have an index on the created_at or updated_at column yet. The updated_at column is new but the created_at existed since the beginning and does not have an index.
- This is not a PR that is supposed to improve performance as the cart didn't have an index on created_at before, I don't intend to add one to change this with this PR. This would need to be a completely different PR and issue.
- This is a background task that shouldn't really need to deal with a lot of data as it is called periodically to delete carts that are too old, so there never should be too many carts in existence.
- If there are too many carts accumulated that need to be deleted, it will still take a while to delete as we do not search for carts here but delete them also, so all corresponding entries need to be deleted also.
- In theory having indices could also slow down a DELETE as it also needs to remove the index references during the DELETE command. Though I admit I would need to run tests in this exact scenario and it probably depends on if more carts are to be deleted than kept or vice versa.

Reply:
Hi @mynameisbogdan thanks for your suggestion.

I adjusted the where clause to ensure indices can be used in the future, but I don't intend to change the restdue to the following points, mostly to keep in synch with Shopware 6's other entities and how the cart table was setup from the start:

- Not intending to set updated_at and created_at together because no other entity in Shopware 6 does so, there really is no sense to break the pattern here.
- I have adjusted the `WHERE` clause so in the future it could take advantage of indices, if necessary. This should be enough in case indices are added later on.
- The `cart` table doesn't have an index on the `created_at` or `updated_at` column yet. The `updated_at` column is new but the created_at existed since the beginning and does not have an index.
- This PR is not supposed to improve the performance, so anything like that should be done in a different PR / issue.
- I actually hope there will never be too many carts because this background task should make sure that old carts are deleted and carts that have been converted to an order are deleted automatically.
- Regarding indices in theory this could actually also slow down a DELETE as the indices also have to be deleted but I guess it depends on the exact scenario and I would need to make some tests. Though it probably is unlikely to take longer but I guess it depends, if more carts have to be deleted or are kept.

Reply:
@mynameisbogdan: Decided to add an index after all as it is just a few lines more code. With the WHERE clause change above it is being used and should speed up the task, if there are a lot of carts.

Reply:
@mynameisbogdan: Decided to add an index after all as it is just a few lines more code. With the WHERE clause change above it is being used and should speed up the task, if there are a lot of carts.

Though I really don't understand the new rule to need to write a migration test. That might make sense for more complex migrations but just for adjusting the scheme it seems kind of pointless because the migration will fail to be executed during test setup anyway. 
And I am not sure complex migrations should be part of the migrations file, those would probably be better in something like a post update core event but OK.

Reply:
Hi @AndreasA,

There's already a index on column `created_at`. Check https://github.com/shopware/platform/blob/trunk/src/Core/Migration/V6_4/Migration1617784658AddCartIndex.php

Reply:
Hi @mynameisbogdan

OK. could be that I didn't know this because I currently only have 6.3.5 installations and there is not index on create.

In any case I have now added a compound index for updated_at and created_at to speed up searches for both. and as updated_at is first, it will also be used in case created_at is not provided, so in combination with created_at everything is covered.

or should I change the index to just an index for updated_at ?

Reply:
IMO since `created_at` already exists and you're not using both columns in a `WHERE` with `AND` I would advise to add a new index only for `updated_at`. 👍

Reply:
@mynameisbogdan I have now created an index just for updated_at which should solve any query issues in combination with the existing index for created_at.

I do a IS NULL / IS NOT NULL in combination with updated_at and created_at but I don't think a compound index will be a huge performance increase to just using two separate indices in such a scenario.

Reply:
That's great. 👍 

You're correct, but it seems that the new index is been used.

![Screenshot 2021-06-15 at 13 07 36](https://user-images.githubusercontent.com/707714/122034960-b4834b00-cdda-11eb-9b4a-0549090e2f0a.jpg)

I use `gedmo/doctrine-extensions` in many of my projects and they set `updated_at` on creation as well, which benefits cases like these where the new query would be simple `DELETE FROM cart WHERE updated_at <= :timestamp`, but I understand that's not the shopware way of doing things. 😊

Reply:
@mynameisbogdan I personally agree with you that setting updatedAt and createdAt together usually makes sense as most of the time one needs to search for a last modified and if I really need to determine only created I can still do a `created_at = updated_at` check but in Shopware 6 all tables use this scheme, so I don't want to break it for one table :smile: